Routing/RouteRegistry.cs

```diff
@@ -1,37 +1,187 @@
+
+        public RouteRegistry(Func<TRequest, TResponse> handleFallbackRequest)
         {
             _handleFallbackRequest = handleFallbackRequest;
         }
 
         public TResponse Route(HttpMethod method, string path, TRequest request)
         {
-            var segments = path.Split('/');
+            var segments = _pathParser.Parse(path)?.ToList();
+            if (segments is null)
+            {
+                return _handleFallbackRequest(request);
+            }
+
+            var requestHandlingData = Match(_segmentTree, method, segments, new Dictionary<string, string>());
+            return requestHandlingData is null
+                ? _handleFallbackRequest(request)
+                : requestHandlingData.HandleRequest(request, requestHandlingData.Parameters);
+        }
+
+        private static RequestHandlingData?
+            Match(SegmentNode<TResponse, TRequest> node,
+                HttpMethod method,
+                ICollection<string> segments,
+                IDictionary<string, string> parameters)
+        {
+            var head = segments.First();
+
+            var currentParameters = node.Matcher is Parameter { Key: var key }
+                ? AddToDictionary(parameters, (key, head))
+                : parameters;
 
-            var routeParams = new Dictionary<string, string> { { "name", "Rubby" }, { "age", "69" } };
+            if (!NodeMatchesSegment(node, head))
+            {
+                return null;
+            }
 
-            return _registeredRoutes[path](request, routeParams);
+            if (segments.Count == 1)
+            {
+                return node.HandleRequestFunctions.TryGetValue(method, out var handleRequest)
+                    ? new RequestHandlingData(handleRequest, currentParameters)
+                    : null;
+            }
+
+            var tail = segments.Skip(1).ToList();
+
+            return node.LiteralChildren
+                       .Concat(node.ParameterChildren)
+                       .Select(child => Match(child, method, tail, currentParameters))
+                       .FirstOrDefault(child => child != null);
+        }
+
+        private static bool NodeMatchesSegment(SegmentNode<TResponse, TRequest> node, string segment) =>
+            !(node.Matcher is Literal { Identifier: var matchingSegment } && segment != matchingSegment);
+
+        private static IDictionary<TKey, TValue> AddToDictionary<TKey, TValue>(IDictionary<TKey, TValue> dictionary, (TKey, TValue) keyValuePair)
+        {
+            var (key, value) = keyValuePair;
+            return new Dictionary<TKey, TValue>(dictionary)
+            {
+                [key] = value
+            };
         }
+
+
         public IRouteRegistry<TResponse, TRequest> Register(HttpMethod method, string route, HandleRequest<TResponse, TRequest> handleRequest)
         {
-            _registeredRoutes[route] = handleRequest;
+            var segments = ParseRoute(route);
```

Comment:
Should we extract the data structure and the insert / remove operations into a separate class?

Reply:
https://github.com/messerli-informatik-ag/routing/pull/6/commits/9190e6d30a507ffc52847bdb0ff7d1075cab4468